### PR TITLE
feat: FractionalIndexUtil for order index

### PR DESCRIPTION
**Pull request by anusham-rc**
### Summary

-

### How to test

1.

### Project Checklist

- [ ] Changes build without any errors
- [ ] Have written adequate test cases
- [ ] Done developer testing in
  - [ ] Chrome
  - [ ] Firefox
  - [ ] Safari
- [ ] Code is formatted with `npm run format`
- [ ] Code is linted with `npm run check-lint`
- [ ] No unnecessary comments left in code
- [ ] Made corresponding changes to the documentation

#### Other

- [ ] New atomic components added
- [ ] New molecules added
- [ ] New pages(routes) added
- [ ] New dependencies installed

### PR Checklist

- [ ] Pull request is raised from the correct source branch
- [ ] Pull request is raised to the correct destination branch
- [ ] Pull request is raised with correct title
- [ ] Pull request is self reviewed
- [ ] Pull request is self assigned
- [ ] Suitable pull request status labels are added (`ready-for-code-review`)

### Additional Information
-

### CODE REVIEW

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        if (a != null) {
+            validateOrderKey(a, digits);
+        }
+        if (b != null) {
+            validateOrderKey(b, digits);
+        }
+        if (a != null && b != null && a.compareTo(b) >= 0) {
+            throw new IllegalArgumentException(a + " >= " + b);
+        }
+
```
[the remark this answers is not in the document]
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java]: no validation needed

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            throw new IllegalArgumentException("invalid order key: " + key);
+        }
+        
+        String intPart = getIntegerPart(key);
+        String fracPart = key.substring(intPart.length());
+        
+        if (!fracPart.isEmpty() && fracPart.substring(fracPart.length() - 1).equals(String.valueOf(digits.charAt(0)))) {
+            throw new IllegalArgumentException("invalid order key: " + key);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java]: no validation needed, keep original

**Review by ThinuwanW — COMMENTED** [review]
## 🤖 Claude Code Review

This PR adds FractionalIndexUtil, a faithful Java port of the canonical rocicorp/fractional-indexing algorithm, plus unit tests. The core algorithm (midpoint, increment/decrement integer, generateKeyBetween, generateNKeysBetween) is correct and matches the reference logic. Findings are minor: a misnamed/ineffective test, a missing guard against negative n that risks StackOverflowError, the util having no production caller yet, and one subtle but safe deviation from the reference in the common-prefix loop.

Found **8** new issue(s): 🔴 1 important, 🟡 1 suggestion(s), 🟣 2 nit(s)

File: backend/src/test/java/com/skapp/community/common/util/FractionalIndexUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+		@Test
+		@DisplayName("a equals b - throws IllegalArgumentException")
+		void generateKeyBetween_AEqualsB_ThrowsException() {
+			String value = "a0";
+			assertThrows(IllegalArgumentException.class, () -> FractionalIndexUtil.generateKeyBetween(value, value));
+		}
+
+		@Test
+		@DisplayName("Invalid key with trailing zero - throws IllegalArgumentException")
+		void generateKeyBetween_InvalidKeyTrailingZero_ThrowsException() {
```
Comment (ThinuwanW) [inline backend/src/test/java/com/skapp/community/common/util/FractionalIndexUtilTest.java:197]: 🤖 **Claude** · 💬 **Note**

This test (and its @DisplayName 'Invalid key with trailing zero') does not test what it claims. generateKeyBetween("a0", "b0") throws because the second argument "b0" has an invalid integer part: head 'b' requires a 3-character integer part (getIntegerLength('b') = 'b'-'a'+2 = 3), so getIntegerPart("b0") throws 'invalid order key'. Note "a0" is itself a perfectly valid key (it's exactly what generateKeyBetween(null,null) returns), so trailing-zero validation is never exercised. There is currently no test covering the actual trailing-zero rejection path in validateOrderKey. Fix: use an input whose fractional part ends in the zero digit, e.g. a key like "a0B0" (integer part "a0", fractional "B0" ending in '0'), and assert it is rejected — then rename/keep this test for the invalid-integer-length case separately.

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+		return ia + midpoint(fa, null, digits);
+	}
+
+	public static List<String> generateNKeysBetween(String a, String b, int n) {
+		return generateNKeysBetween(a, b, n, BASE_62_DIGITS);
+	}
+
+	public static List<String> generateNKeysBetween(String a, String b, int n, String digits) {
+		if (n == 0) {
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:255]: 🤖 **Claude** · 💬 **Note**

generateNKeysBetween has no guard for negative n. With both a and b non-null and n < 0, mid = floor(n/2.0) stays negative and the recursive call generateNKeysBetween(a, c, mid, digits) recurses indefinitely with a shrinking interval but a never-terminating negative n, leading to StackOverflowError rather than a clean error. (For the b==null / a==null branches a negative n silently returns a single key instead of failing.) Suggest adding a fail-fast check at the top, e.g. `if (n < 0) throw new IllegalArgumentException("n must be >= 0");`, so caller misuse surfaces as a clear IllegalArgumentException.

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
+package com.skapp.community.common.util;
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:1]: 🤖 **Claude** · 💬 **Note**

📁 **File-level observation**

This utility (and its full test suite) is introduced but has no production caller — grep shows FractionalIndexUtil is referenced only by its own test. If it's foundational for an upcoming ordering/reordering feature that's fine, but otherwise it will sit as dead code. Confirm a follow-up PR will wire it in, or hold it until the consuming feature lands.

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+		if (!a.isEmpty() && a.charAt(a.length() - 1) == zero
+				|| (b != null && !b.isEmpty() && b.charAt(b.length() - 1) == zero)) {
+			throw new IllegalArgumentException("trailing zero");
+		}
+
+		if (b != null) {
+			int n = 0;
+			while (n < Math.max(a.length(), b.length())) {
+				char charA = n < a.length() ? a.charAt(n) : zero;
+				char charB = n < b.length() ? b.charAt(n) : zero;
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:30]: 🤖 **Claude** · 💬 **Note**

Nit (safe, but worth a comment): the common-prefix loop deviates from the reference algorithm by padding b with the zero digit (charB = zero when n >= b.length()), whereas the canonical implementation compares against the raw character of b and stops when b is exhausted. This is provably equivalent here only because midpoint already rejects any b ending in the zero digit (the trailing-zero check above) and a < b is guaranteed, so the loop can never advance past b.length() with all characters matching. Since correctness relies on that non-obvious invariant, consider a brief comment, or mirror the reference's unpadded b comparison to make future upstream fixes easier to apply.

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+				return b.substring(0, 1);
+			}
+			else {
+				return digits.charAt(digitA) + midpoint(a.substring(1), null, digits);
+			}
+		}
+	}
+
+	private static void validateInteger(String intPart) {
+		if (intPart.length() != getIntegerLength(intPart.charAt(0))) {
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:58]: 🤖 **Claude** · 🔴 **Important**

Latent crash / behavioral drift from the reference algorithm. When midpoint is reached with a="" and a single-digit b whose first digit is consecutive to digitA (e.g. b="1"), this branch executes `midpoint(a.substring(1), null, digits)` with a empty. In Java, "".substring(1) throws StringIndexOutOfBoundsException (the JS reference's "".slice(1) returns "" and proceeds, returning a valid key). This is reachable from valid input: generateKeyBetween(null, b) where b is at the bottom-of-space boundary integer part "A"+26 zeros followed by a single fractional digit (e.g. b="A00000000000000000000000001"). That key passes validateOrderKey, hits the `ib.equals("A"+repeated)` branch (line 213-214) which calls midpoint("", "1", digits), then crashes here. Fix: guard the recursion as the reference does — when a is empty, treat a.substring(1) as "" (e.g. `String rest = a.isEmpty() ? "" : a.substring(1); return digits.charAt(digitA) + midpoint(rest, null, digits);`). Recommend adding a unit test for generateKeyBetween(null, "A"+"0".repeat(26)+"1").

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+	}
+
+	private static String getIntegerPart(String key) {
+		int integerPartLength = getIntegerLength(key.charAt(0));
+		if (integerPartLength > key.length()) {
+			throw new IllegalArgumentException("invalid order key: " + key);
+		}
+		return key.substring(0, integerPartLength);
+	}
+
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:82]: 🤖 **Claude** · 🟣 **Nit**

Inconsistent exception type for empty-string input. The class otherwise reports malformed keys via IllegalArgumentException ("invalid order key..."), but an empty string slips past that contract: generateKeyBetween("", b) -> validateOrderKey -> getIntegerPart -> getIntegerLength(key.charAt(0)) calls "".charAt(0), throwing StringIndexOutOfBoundsException instead of IllegalArgumentException. Callers catching IllegalArgumentException to handle bad keys will leak an unexpected exception type. Fix: validate non-empty key at the entry of getIntegerPart/validateOrderKey (or in generateKeyBetween) and throw IllegalArgumentException with the standard "invalid order key" message.

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java (hunk 5 above)
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:1]: 🤖 **Claude** · 🟡 **Suggestion**

📁 **File-level observation**

Missing documentation and source attribution. This is a direct port of the widely-used `fractional-indexing` algorithm (MIT-licensed, by David Greenspan / Observable's "Implementing Fractional Indexing"). There is no class-level Javadoc describing the key format (integer part + fractional part, base-62), the ordering contract, the exception contract, or thread-safety, and no attribution/license note. Given the non-obvious algorithm, add a class-level Javadoc explaining the format and linking the source, plus brief Javadoc on the two public generateKeyBetween/generateNKeysBetween entry points documenting null semantics and the IllegalArgumentException conditions.

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+		return generateKeyBetween(a, b, BASE_62_DIGITS);
+	}
+
+	public static String generateKeyBetween(String a, String b, String digits) {
+		if (a != null) {
+			validateOrderKey(a, digits);
+		}
+		if (b != null) {
+			validateOrderKey(b, digits);
+		}
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:193]: 🤖 **Claude** · 🟣 **Nit**

Public custom-`digits` overloads accept an unvalidated digit set. generateKeyBetween(a,b,digits) and generateNKeysBetween(a,b,n,digits) are public, but a caller passing an empty string (digits.charAt(0) -> StringIndexOutOfBoundsException), a single-character set, or a set with duplicate/unsorted characters will get crashes or silently incorrect ordering, since the algorithm assumes digits are distinct and in ascending order. Consider either reducing these overloads to package-private (only the BASE_62 public API is needed by callers) or validating `digits` (non-empty, distinct, sorted) at entry.

File: backend/src/test/java/com/skapp/community/common/util/FractionalIndexUtilTest.java (hunk 3 above)
Comment (anusham-rc) [reply backend/src/test/java/com/skapp/community/common/util/FractionalIndexUtilTest.java:197]: added required tests

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java (hunk 4 above)
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:255]: Added the guard to the 4-arg overload (the one all paths funnel into). The 3-arg overload delegates to it unchanged, so negative n now throws IllegalArgumentException immediately regardless of which overload is called.

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java (hunk 5 above)
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:1]: Need to discuss with the team

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java (hunk 6 above)
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:30]: no changes added, keeping the original code

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java (hunk 7 above)
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:58]: no changes added, keeping the original code

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java (hunk 8 above)
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:82]: no changes added, keeping the original code

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java (hunk 5 above)
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:1]: Not the practice

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java (hunk 9 above)
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:193]: no changes added, keeping the original code

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        if ((a.isEmpty() ? "" : a.substring(a.length() - 1)).equals(String.valueOf(zero)) ||
+            (b != null && (b.isEmpty() ? "" : b.substring(b.length() - 1)).equals(String.valueOf(zero)))) {
+            throw new IllegalArgumentException("trailing zero");
+        }
+        
+        if (b != null) {
+            int n = 0;
+            while (n < a.length() && n < b.length() && 
+                   (n >= a.length() ? zero : a.charAt(n)) == (n >= b.length() ? zero : b.charAt(n))) {
+                n++;
```
[the remark this answers is not in the document]
Comment (anusham-rc) [reply backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java]: no changes needed, keep original

**Review by akilarootcode — APPROVED** [review]

File: backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+		int mid = (int) Math.floor(n / 2.0);
+		String c = generateKeyBetween(a, b, digits);
+		List<String> result = new ArrayList<>(generateNKeysBetween(a, c, mid, digits));
+		result.add(c);
+		result.addAll(generateNKeysBetween(c, b, n - mid - 1, digits));
+		return result;
+	}
+
+}
```
Comment (ThinuwanW) [inline backend/src/main/java/com/skapp/community/common/util/FractionalIndexUtil.java:295]: verify the improvements from comments with Yuhanga

**Review by ThinuwanW — CHANGES_REQUESTED** [review]
